### PR TITLE
Check if a graph is a forest

### PR DESCRIPTION
**Pull request by rohitt28**

### CODE REVIEW

**Comment by rohitt28** [comment]
Initial draft to implement the igraph_is_graph function mentioned [here](https://github.com/igraph/igraph/issues/1873)

**Comment by GroteGnoom** [comment]
Don't forget to add your function to `doc/structural.xxml`, so it's documentation will be included.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:82]: Right now you don't destroy this one.

Comment (rohitt28) [reply src/properties/forest.c:82]: Thanks for pointing it out. I will change that in next commit.

**Comment by GroteGnoom** [comment]
I think you've got a lot of the structure already correctly figured out 🙂
You've not added documentation yet, you can already include that as well.
I think you can continue. I'm interested to see your algorithm 🙂

File: tests/unit/igraph_is_forest.c
Comment (GroteGnoom) [inline tests/unit/igraph_is_forest.c:1]: You're missing the copyright here.

**Comment by rohitt28** [comment]
Hey @GroteGnoom, I tried to run unit tests locally and realized that ```igraph_is_forest``` function isn't recognised in the files. I thought adding it to ```igraph_structural.h``` would be enough. Can you help me fix this?

**Comment by rohitt28** [comment]
The error when I run the tests ```168/401 Test #168: test::igraph_is_forest.............................***Failed    0.70 sec
Test exited abnormally with error: Access violation```

And when I hover over the function it shows this instead of the parameters

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:27]: use igraph_integer_t instead of longs.

**Comment by rohitt28** [comment]
This is the current version, it works when I add the function on top of it.

**Comment by GroteGnoom** [comment]
For me it builds fine but segfaults. I'll check things out and add some comments.

File: tests/unit/igraph_is_forest.c
Comment (GroteGnoom) [inline tests/unit/igraph_is_forest.c:24]: You don't have to put things on the finally stack in tests.

File: tests/unit/igraph_is_forest.c
Comment (GroteGnoom) [inline tests/unit/igraph_is_forest.c:37]: If you print something before each test you can see which output belongs to which test. Else the test results are hard to check.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:211]: You do a CLEAN of 3, but you only destroy 2 things.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:195]: You don't FINALLY_CLEAN this one.

**Comment by GroteGnoom** [comment]
Your undirected graph with cycle seems to segfault.

**Comment by GroteGnoom** [comment]
As for your build problems: Did you do a `cmake ..` and then a `cmake --build . --target check` ?

**Comment by rohitt28** [comment]
> Your undirected graph with cycle seems to segfault.

Oh! I accidentally added 6 vertices instead of 5. 


> Did you do a `cmake ..` and then a `cmake --build . --target check`

Yes, I will try to clone the repo again in another folder, maybe that might work.

**Comment by rohitt28** [comment]
I have made the required changes, could you check if I missed something. Also currently this function is in `forest.c` (new file) but it could be moved to the `trees.c` file. Let me know if moving it there would be better.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:3]: I think the copyright headers should have like 3 extra newlines in them, which should be the case in most other files.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:24]: We don't do ints anymore, so this int should be an igraph_integer_t

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:27]: Can you put a newline between your declarations and your code? That makes it a little easier to read.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:87]: Isn't there something missing from this definition? It seems to allow for a vertex with two parents. Like 0->1, 0->2, 1->3, 2->3 should not be in forest, right? I see that it's also in `is_tree`, but I'm still confused.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:115]: The return type should be igraph_error_t

File: src/properties/forest.c
Comment (rohitt28) [reply src/properties/forest.c:87]: Oh thats right. That condition is checked in the code, I will add it in the description too.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:109]: This is actually an IGRAPH_EINVMODE

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:203]: Can you give me an example of when then visited_count is not the vcount? I think that shouldn't happen for this algorithm.

Comment (rohitt28) [reply src/properties/forest.c:203]: If there is a single vertex in directed graph with a self loop then its degree wont be 0 so `igraph_forest_visitor` wont be called for that vertex and function would return that it is a forest.

Comment (GroteGnoom) [reply src/properties/forest.c:203]: Aha! Thanks!

**Comment by GroteGnoom** [comment]
There's also another issue that needs to be resolved: the conflict with the include/igraph_structural.h
Did you maybe base your branch on master instead of develop?

**Comment by rohitt28** [comment]
>  Did you maybe base your branch on master instead of develop?

Yes, thats why the functions still had int return type. I think that merge conflict can be resolved by just removing the int functions. Will this work?

**Comment by GroteGnoom** [comment]
I think so, it seems like it's going to work out.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:41]: also here some space between declarations and code please :)

**Comment by GroteGnoom** [comment]
All right, that merge cleaned a lot of things up, thanks :) I think I will stop reviewing soon, and maybe continue tomorrow. We should try to pass the tests, but right now your `res` is of different types in the two functions, so it will not compile.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:153]: You can init this with vcount immediately, so you don't have to resize it later. All the elements are set to 0 on init, so you also can remove the loop.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:125]: You assume the roots vector starts with  a size of 0 later in the algorithm, but that doesn't always hold, so you need to resize it yourself.

File: src/properties/forest.c
Comment (GroteGnoom) [inline src/properties/forest.c:155]: This resize should be moved up higher, because at line 142 you already do         igraph_vector_push_back(roots,0); for `vcount == 1`

Comment (rohitt28) [reply src/properties/forest.c]: Done

**Comment by GroteGnoom** [comment]
I think looks pretty great, and you done very good work. It was a pleasure working with you on this, thanks 🙂 I will leave it up to someone else do the final checks. @ntamas, can you take a look?

**Comment by rohitt28** [comment]
Thanks a lot for your helpful reviews. I will keep these suggestions in mind while contributing again.

File: include/igraph_structural.h
Comment (ntamas) [inline include/igraph_structural.h:63]: This empty line is unnecessary here.

File: include/igraph_structural.h
Comment (ntamas) [inline include/igraph_structural.h:57]: This empty line is unnecessary here.

File: doc/structural.xxml
Comment (ntamas) [inline doc/structural.xxml:153]: I'd rather put the `igraph_is_forest()` function in the Trees section for the time being. I don't know how many forest-related functions we are going to have, but forests and trees are closely related anyway so we can keep them in the same section. Just rename the section to "Trees and forests"

File: src/properties/forest.c
Comment (ntamas) [inline src/properties/forest.c:27]: Please wrap long lines at around 80 chars or so to make it easier to read on narrow screens or for people who use two code panels side-by-side.

File: src/properties/forest.c
Comment (ntamas) [inline src/properties/forest.c:120]: This is my only real concern with the current state of this PR: the `roots` argument should be of type `igraph_vector_int_t`, _not_ `igraph_vector_t`. `igraph_vector_t` contains doubles, not integers.

Also, nitpicking: please include a space after every comma, and no space before commas.

File: src/properties/forest.c
Comment (ntamas) [inline src/properties/forest.c:143]: Please add an `IGRAPH_CHECK()` here. It is very unlikely that the user won't have free memory for pushing one more item to the roots vector, but it may still happen and it's best if you get into the habit of wrapping all igraph function calls in `IGRAPH_CHECK()`.

File: src/properties/forest.c
Comment (ntamas) [inline src/properties/forest.c:175]: `IGRAPH_CHECK()` is missing here.

File: src/properties/forest.c
Comment (ntamas) [inline src/properties/forest.c:187]: @szhorvat has recently introduced a macro named `IGRAPH_REVERSE_MODE(...)`, which reverses the `mode` argument like you do it here. Let's start using it. This is not in the docs yet so that's why I'm telling now :)

File: src/properties/forest.c
Comment (ntamas) [inline src/properties/forest.c:171]: In the rest of the codebase, `case` branches are indented by one indentation unit (4 spaces) relative to the `switch` statement. Please update the coding style accordingly.

File: src/properties/forest.c
Comment (ntamas) [inline src/properties/forest.c:150]: I would not use an adjacency list for this algorithm, I'd simply use `igraph_neighbors()` whenever I need the neighbors of a vertex. The reason is that basically `igraph_adjlist_init()` is pre-querying `igraph_neighbors()` for all the vertices, and this is wasteful if you are likely to visit every vertex at most once. We typically only use `igraph_adjlist_init()` where we know that `igraph_neighbors()` queries will be repeated over and over multiple times for many vertices, otherwise there is no performance gain in using `igraph_adjlist_init()` but there is an additional cost you pay in storage.

**Review by ntamas — CHANGES_REQUESTED** [review]
@rohitt28 Thank you for your work on this PR, great job! I've added a few comments; the two major concerns that I raised are:

* the `roots` argument should be an `igraph_vector_int_t`, not an `igraph_vector_t` because the latter stores doubles
* there is no need to use an adjacency list because you will visit each vertex at most once. `igraph_adjlist_init()` is doing an `igraph_neighbors()` query for each vertex in the graph in advance, which pays off performance-wise only if you would visit the same vertex multiple times. This is not the case here.

@GroteGnoom Thanks for the initial round of reviews! I'll take over from here until this PR is merged.

**Comment by ntamas** [comment]
Also, please take a look at the code coverage report [here](https://app.codecov.io/gh/igraph/igraph/compare/1888/diff) and try to exercise all newly added code paths in the unit tests. It seems like the case of the single-vertex graph is not covered. There is a branch in the forest visitor that the unit tests do not cover either.

**Comment by szhorvat** [comment]
I'd like to have the chance to review this before it is merged. Please give me a couple of days.

File: doc/structural.xxml
Comment (szhorvat) [inline doc/structural.xxml:144]: "forests" should be plural.

"forests" should not be capitalized: let's keep the style consistent with the rest of this document.

File: src/properties/trees.c
Comment (szhorvat) [inline src/properties/trees.c:459]: Let us use `igraph_vector_int_clear()` here and drop `IGRAPH_CHECK` (because `igraph_vector_int_clear()` never fails and thus does not return an error code).

Comment (ntamas) [inline src/properties/trees.c:459]: `igraph_is_forest()` should allow the user to pass in a null pointer for `roots` if the user is not interested in the roots and only wants a yes/no answer in `res`. it should also allow `res` to be null if the user is only interested in the roots. Basically you would need to guard accesses to `*res` and `*roots` with `if (...)` statements.

**Review by szhorvat — CHANGES_REQUESTED** [review]
I left a few comments, but most of these are nitpicks that are easily fixed, and they are not what I'm worried about.

My major concern is that I am not convinced that the algorithm is correct, i.e. that cycles will always be detected correctly. Specifically, in the undirected case, only _unvisited_ neighbours of a vertex are pushed onto the stack. Visited ones are skipped. This is done to avoid traversing the graph backwards towards a vertex's parent in the DFS tree. However, the check you do is not specifically for the parent, but for any visited vertex.

Could it happen that a vertex will not be pushed onto the stack due to this check, preventing the detection of a real cycle?  Can you prove that this can never happen?  It may be true, but it is far from obvious to me.

I don't have time to dig into deeper today, so I'll stop here for now.

I will also note that counting visited vertices and checking that all of them were reached seems pointless here. For `is_tree` we did it to ensure that the graph was connected: unlike forests, all trees are connected. For forests, this is not necessary.

**Comment by szhorvat** [comment]
Actually I think this is correct. The question can be put as: if there is a cycle in the graph, is it certain that at least one vertex will be "discovered" more than once (i.e. pushed onto the stack more than once) before it is fully processed.

The remaining part is the counting of visited vertices: why did you keep this check?

**Comment by rohitt28** [comment]
> counting of visited vertices: why did you keep this check?

Do you mean this one?
``if(*res)
{
          *res= visited_count==vcount;
}``

If yes then, in case of directed graph there could be 1 separate component with only 1 node that has self loop.
That node will have degree 1 so the  ``igraph_forest_visitor`` function will not be called for it. Then if rest of graph is a forest then we will get wrong output.

**Comment by rohitt28** [comment]
>  if there is a cycle in the graph, is it certain that at least one vertex will be "discovered" more than once (i.e. pushed onto the stack more than once) before it is fully processed.

You are probably right about that. I tried some test cases in which I thought current algo will fail but it seems to pass.
For eg: `igraph_small(&graph, 6, 0, 0, 1, 0, 2, 2, 1, -1);` Anyways, I will just use another vector to keep track of parents.

**Comment by szhorvat** [comment]
> If yes then, in case of directed graph there could be 1 separate component with only 1 node that has self loop.
> That node will have degree 1 so the `igraph_forest_visitor` function will not be called for it. Then if rest of graph is a forest then we will get wrong output.

Yes, you are correct about this.

Please give me a couple more days to convince myself about the correctness of the algorithm, or otherwise find a counterexample, and I'll get back to you. One reason why it is tricky is that if we skip pushing visited vertices on the stack then unvisited vertices will be pushed more than once only in the undirected case, not the directed case. That's of course not a problem because we only check for the visited status in the undirected case. I'm just explaining why I found this confusing: even though the code for the directed and undirected cases is almost identical, the way they work is quite different.

**Comment by szhorvat** [comment]
@rohitt28 Thanks again for implementing this! Everything's fine with the algorithm. Can you fix the small issues I commented on, and double-check the formatting? Then I'll add a few comments and a few small documentation edits and we can merge this.

**Comment by szhorvat** [comment]
Summary of changes I made:

There were a couple of places where `roots` was accessed without checking that it is not NULL.

We now re-use the stack and neighbours vector across calls to the visitor function.

Added another test.

The rest are cosmetic changes.

**Comment by szhorvat** [comment]
I think this is good to merge now.

Thanks so much for your work @rohitt28!

**Review by ntamas — APPROVED** [review]

**Comment by ntamas** [comment]
Thanks a lot @rohitt28 !

**Pull request merged**
